src/devices/cpu/tlcs900/tmp94c241.cpp

```diff
@@ -0,0 +1,1769 @@
+
+	switch(operation)
+	{
+	case FF_INVERT:
+		*ff_state = !(*ff_state);
+		break;
+	case FF_SET:
+		*ff_state = true;
+		break;
+	case FF_CLEAR:
+		*ff_state = false;
+		break;
+	default:
+		// invalid operation
+		return;
+	}
+
+	/* The value of the flipflop is only exposed to a pin in certain modes of operation
+	   determined by fields of the port function registers.
+
+	   So here we bail out if the flipflop is not routed to its corresponding port bit:
+	*/
+	switch( flipflop )
+		{
+		case 0x1:
+			if (!BIT(m_port_function[PORT_C], 0) || BIT(m_port_control[PORT_C], 0)) return;
+			break;
+		case 0x7:
+			if (!BIT(m_port_function[PORT_C], 0) || !BIT(m_port_control[PORT_C], 0)) return;
+			break;
+		case 0x3:
+			if (!BIT(m_port_function[PORT_C], 1) || BIT(m_port_control[PORT_C], 1)) return;
+			break;
+		case 0xb:
+			if (!BIT(m_port_function[PORT_C], 1) || !BIT(m_port_control[PORT_C], 1)) return;
+			break;
+		case 0x4:
+			if (!BIT(m_port_function[PORT_D], 0)) return;
+			break;
+		case 0x6:
+			if (!BIT(m_port_function[PORT_D], 4)) return;
+			break;
+		case 0x8:
+			if (!BIT(m_port_function[PORT_E], 0)) return;
+			break;
+		case 0xA:
+			if (!BIT(m_port_function[PORT_E], 4)) return;
+			break;
+		default:
+			// invalid flip flop
+			return;
+		}
+
+	/* And here we actually send the value to the corresponding pin */
+	uint8_t new_port_value = 0;
+	switch( flipflop )
+		{
+		case 0x1:
+		case 0x7:
+			new_port_value = m_port_latch[PORT_C] & 0xfe;
+			if (*ff_state) new_port_value |= 0x01;
+			port_w<PORT_C>(new_port_value);
+			break;
+		case 0x3:
+		case 0xb:
+			new_port_value = m_port_latch[PORT_C] & 0xfd;
+			if (*ff_state) new_port_value |= 0x02;
+			port_w<PORT_C>(new_port_value);
+			break;
+		case 0x4:
+			new_port_value = m_port_latch[PORT_D] & 0xfe;
+			if (*ff_state) new_port_value |= 0x01;
+			port_w<PORT_D>(new_port_value);
+			break;
+		case 0x6:
+			new_port_value = m_port_latch[PORT_D] & 0xef;
+			if (*ff_state) new_port_value |= 0x10;
+			port_w<PORT_D>(new_port_value);
+			break;
+		case 0x8:
+			new_port_value = m_port_latch[PORT_E] & 0xfe;
+			if (*ff_state) new_port_value |= 0x01;
+			port_w<PORT_E>(new_port_value);
+			break;
+		case 0xa:
+			new_port_value = m_port_latch[PORT_E] & 0xef;
+			if (*ff_state) new_port_value |= 0x10;
+			port_w<PORT_E>(new_port_value);
+			break;
+		}
+}
+
+void tmp94c241_device::tffcr_w(uint8_t data)
+{
+	change_timer_flipflop( 1, (data >> 2) & 3 );
+	change_timer_flipflop( 3, (data >> 6) & 3 );
+	m_tffcr = data | 0xcc;
+}
+
+void tmp94c241_device::treg23_w(offs_t offset, uint8_t data)
+{
+	if (BIT(offset, 0))
+		m_treg3 = data;
+	else
+		m_treg2 = data;
+}
+
+uint8_t tmp94c241_device::t23mod_r()
+{
+	return m_t23mod;
+}
+
+void tmp94c241_device::t23mod_w(uint8_t data)
+{
+	m_t23mod = data;
+}
+
+uint8_t tmp94c241_device::trdc_r()
+{
+	return m_trdc;
+}
+
+void tmp94c241_device::trdc_w(uint8_t data)
+{
+	m_trdc = data;
+}
+
+void tmp94c241_device::treg45_w(offs_t offset, uint16_t data)
```

Comment:
Obviously untested code is bad, mmmkay?  offset is going to be 0 or 1, never 2.

Reply:
`map(0x000090, 0x000093).w(FUNC(tmp94c241_device::treg45_w));`

I expect offset to be between 0 and 3.

Reply:
oh, I see! data is `uint16_t`... ouch!

Reply:
I'll make additional commits later tonight. Thanks for finding this bug!

Reply:
I think I fixed it now.